From c62586a88a7e53a43c9a129eb45625c45c6e33c1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20Le=20Goffic?= <clement.legoffic@foss.st.com>
Date: Mon, 2 Sep 2024 11:06:36 +0200
Subject: [PATCH 1144/1695] core: pta: verify RTC PTA data buffer is aligned
 with struct pta_rtc_time
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Verify in function rtc_pta_get_time() and rtc_pta_set_time() that
data buffers are aligned with struct pta_rtc_time.
Also introduce rtc_pta_copy_time_to_optee() to copy a
struct pta_rtc_time in a struct optee_rtc_time.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git cc47b89f7d36f860e59ec674efbab6e4cb03f398 core: pta: verify RTC PTA data buffer is aligned with struct pta_rtc_time]
Change-Id: I19499a2c5a48ea06dd8ecf88c4744de01efecd03
Signed-off-by: Clément Le Goffic <clement.legoffic@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/402358
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/pta/rtc.c | 50 ++++++++++++++++++++++++++++----------------------
 1 file changed, 28 insertions(+), 22 deletions(-)

diff --git a/core/pta/rtc.c b/core/pta/rtc.c
index de2eade56..ff69319d6 100644
--- a/core/pta/rtc.c
+++ b/core/pta/rtc.c
@@ -27,6 +27,18 @@ static void rtc_pta_copy_time_from_optee(struct pta_rtc_time *pta_time,
 	pta_time->tm_wday = optee_time->tm_wday;
 }
 
+static void rtc_pta_copy_time_to_optee(struct optee_rtc_time *optee_time,
+				       struct pta_rtc_time *pta_time)
+{
+	optee_time->tm_sec = pta_time->tm_sec;
+	optee_time->tm_min = pta_time->tm_min;
+	optee_time->tm_hour = pta_time->tm_hour;
+	optee_time->tm_mday = pta_time->tm_mday;
+	optee_time->tm_mon = pta_time->tm_mon;
+	optee_time->tm_year = pta_time->tm_year;
+	optee_time->tm_wday = pta_time->tm_wday;
+}
+
 static TEE_Result rtc_pta_get_time(uint32_t types,
 				   TEE_Param params[TEE_NUM_PARAMS])
 {
@@ -41,16 +53,15 @@ static TEE_Result rtc_pta_get_time(uint32_t types,
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	pta_time = params[0].memref.buffer;
-	if (!pta_time || params[0].memref.size != sizeof(*pta_time))
+	if (!pta_time || params[0].memref.size != sizeof(*pta_time) ||
+	    !IS_ALIGNED_WITH_TYPE(pta_time, typeof(*pta_time)))
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	res = rtc_get_time(&time);
-	if (res)
-		return res;
-
-	rtc_pta_copy_time_from_optee(pta_time, &time);
+	if (!res)
+		rtc_pta_copy_time_from_optee(pta_time, &time);
 
-	return TEE_SUCCESS;
+	return res;
 }
 
 static TEE_Result rtc_pta_set_time(uint32_t types,
@@ -66,16 +77,11 @@ static TEE_Result rtc_pta_set_time(uint32_t types,
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	pta_time = params[0].memref.buffer;
-	if (!pta_time || params[0].memref.size != sizeof(*pta_time))
+	if (!pta_time || params[0].memref.size != sizeof(*pta_time) ||
+	    !IS_ALIGNED_WITH_TYPE(pta_time, typeof(*pta_time)))
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	time.tm_sec = pta_time->tm_sec;
-	time.tm_min = pta_time->tm_min;
-	time.tm_hour = pta_time->tm_hour;
-	time.tm_mday = pta_time->tm_mday;
-	time.tm_mon = pta_time->tm_mon;
-	time.tm_year = pta_time->tm_year;
-	time.tm_wday = pta_time->tm_wday;
+	rtc_pta_copy_time_to_optee(&time, pta_time);
 
 	return rtc_set_time(&time);
 }
@@ -105,13 +111,12 @@ static TEE_Result rtc_pta_get_offset(uint32_t types,
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	res = rtc_get_offset(&offset);
-	if (res)
-		return res;
-
-	if (offset > INT32_MAX || offset < INT32_MIN)
-		return TEE_ERROR_OVERFLOW;
-
-	params[0].value.a = (uint32_t)offset;
+	if (!res) {
+		if (offset > INT32_MAX || offset < INT32_MIN)
+			res = TEE_ERROR_OVERFLOW;
+		else
+			params[0].value.a = (uint32_t)offset;
+	}
 
 	return res;
 }
@@ -132,7 +137,8 @@ static TEE_Result rtc_pta_get_info(uint32_t types,
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	info = params[0].memref.buffer;
-	if (!info || params[0].memref.size != sizeof(*info))
+	if (!info || params[0].memref.size != sizeof(*info) ||
+	    !IS_ALIGNED_WITH_TYPE(info, typeof(*info)))
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	memset(info, 0, sizeof(*info));
-- 
2.47.2

